Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -218,8 +221,31 @@ void TextureMapperLayer::paintSelf(TextureMapperPaintOptions& options)
         solidColorLayer.setColor(m_state.solidColor);
         contentsLayer = &solidColorLayer;
     }
-    if (!contentsLayer)
+
+    if (!contentsLayer) {
+        // Use the damage information we received from the CoordinatedGraphicsLayer
+        // Here we ignore the targetRect parameter as it should already have
+        // been covered by the damage tracking in setNeedsDisplay/setNeedsDisplayInRect
+        // calls from CoordinatedGraphicsLayer.
+        if (m_recordDamage) {
+            if (m_damage.isInvalid())
+                recordDamage(layerRect(), transform, options);
```

Comment:
Whenenver we find a situation in which we need to damage the *whole* frame (and not only the current layer!) then we mark a `Damage` object as invalid. Then when damage is propagated to the scene `Damage` any invalid layer in the tree will cause the scene one to also be marked invalid. At the point where we mark layer damage as invalid, we may not know the exact extents because a layer that is 3D-transformed may not cover the area corresponding to its size!

The way I like to think about it is: if any `Damage` is marked invalid, it will “taint” and damage any other `Damage` where it gets combined, resulting in the full frame considered damaged.

Hopefully we may be able to remove the concept of invalidation and do as you suggest in all cases in the future. For example this patch already does what you suggest for non-transformed content layers: the whole layer rect is damaged instead.

Should I add a comment about how this works to the `Damage` class source?